From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ronak Pithadiya <rpithadiya@nvidia.com>
Date: Wed, 13 Aug 2025 07:31:03 +0000
Subject: [PATCH 094/243] NVIDIA: SAUCE: tegra_hv: Add sysfs to trigger HVC

Added hyp_guest_enter_vm_op function to trigger
hypercall to update VM_OP transition completed
to sysmgr, and exposed sysfs entry
(update_vm_op) to trigger it from userspace.

nvdvms_server will use the sysfs entry to
inform sysmgr about VM_OP transition completion.

Bug 3977687

Change-Id: I97a20d865e715be607a9eaa7d8c305caa0507f84
Signed-off-by: Ronak Pithadiya <rpithadiya@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3429867
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
Reviewed-by: Deepak Nibade <dnibade@nvidia.com>
(cherry picked from commit 451f77c9111fec082663da4cd1238bc343b18257)

diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index 8de28870da846..46f259e1cc468 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -29,6 +29,11 @@
 #define INFO(...) pr_info("tegra_hv: " __VA_ARGS__)
 #define DRV_NAME	"tegra_hv"
 
+/* NVDVMS VM operation constants */
+#define NVDVMS_VM_OP_MIN 300U
+#define NVDVMS_VM_OP_MAX 363U
+#define ENABLE_PM_SERVER_DT_NODE "enable_pm_server"
+
 struct tegra_hv_data;
 
 static struct property interrupts_prop = {
@@ -404,6 +409,53 @@ static ssize_t vmid_show(const struct class *class,
 }
 static CLASS_ATTR_RO(vmid);
 
+static bool is_device_node_present(const char *node_name)
+{
+	struct device_node *node;
+	bool node_present = false;
+
+	node = of_find_node_by_name(NULL, node_name);
+	if (node) {
+		node_present = true;
+		of_node_put(node);
+		INFO("Found %s node in device tree\n", node_name);
+	} else {
+		INFO("No %s node found in device tree\n", node_name);
+	}
+
+	return node_present;
+}
+
+
+static ssize_t update_vm_op_store(const struct class *class,
+		const struct class_attribute *attr, const char *buf, size_t count)
+{
+	unsigned int state;
+	int ret;
+
+	/* Parse unsigned int from user input */
+	ret = kstrtouint(buf, 0, &state);
+	if (ret < 0)
+		return ret;
+
+	/* Validate state is within VM operation boundaries */
+	if (state < NVDVMS_VM_OP_MIN || state > NVDVMS_VM_OP_MAX) {
+		pr_err("tegra_hv: Invalid VM_OP state %u\n", state);
+		return -EINVAL;
+	}
+
+	/* Trigger HVC to update VM_OP transition completion */
+	ret = hyp_guest_enter_vm_op(GUEST_ENTER_VM_OP_CMD, state);
+	if (ret < 0) {
+		pr_err("tegra_hv: VM_OP transition update failed\n");
+		return ret;
+	}
+
+	pr_info("tegra_hv: VM_OP transition updated: %u\n", state);
+	return count;
+}
+static CLASS_ATTR_WO(update_vm_op);
+
 static int tegra_hv_setup(struct tegra_hv_data *hvd)
 {
 	const int intr_property_size = 3;
@@ -438,6 +490,15 @@ static int tegra_hv_setup(struct tegra_hv_data *hvd)
 		return ret;
 	}
 
+	/* Create sysfs entry only if enable_pm_server DT node is present */
+	if (is_device_node_present(ENABLE_PM_SERVER_DT_NODE)) {
+		ret = class_create_file(hvd->hv_class, &class_attr_update_vm_op);
+		if (ret != 0) {
+			ERR("failed to create update_vm_op file: %d\n", ret);
+			return ret;
+		}
+	}
+
 	ret = hyp_read_ivc_info(&info_page);
 	if (ret != 0) {
 		ERR("failed to obtain IVC info page: %d\n", ret);
diff --git a/include/soc/tegra/virt/syscalls.h b/include/soc/tegra/virt/syscalls.h
index 18fb0bb8e582e..c33a01fc6714c 100644
--- a/include/soc/tegra/virt/syscalls.h
+++ b/include/soc/tegra/virt/syscalls.h
@@ -376,6 +376,20 @@ static inline int hyp_guest_reset(unsigned int id, struct hyp_sys_state_info *ou
 	return (int)r0;
 }
 
+__attribute__((no_sanitize_address))
+static inline int hyp_guest_enter_vm_op(unsigned int const id, unsigned int const vm_op)
+{
+	register uint64_t r0 asm("x0") = id;
+	register uint64_t r1 asm("x1") = vm_op;
+
+	asm volatile("hvc %2"
+		: "+r"(r0), "+r"(r1)
+		: "i"(HVC_NR_GUEST_RESET)
+		: "x2", _X3_X17);
+
+	return (int)r0;
+}
+
 __attribute__((no_sanitize_address))
 static inline uint64_t hyp_sysinfo_ipa(void)
 {
diff --git a/include/soc/tegra/virt/tegra_hv_sysmgr.h b/include/soc/tegra/virt/tegra_hv_sysmgr.h
index e735b6f172495..d8ac0cc255521 100644
--- a/include/soc/tegra/virt/tegra_hv_sysmgr.h
+++ b/include/soc/tegra/virt/tegra_hv_sysmgr.h
@@ -93,6 +93,7 @@ enum system_function_id {
 	GUEST_PAUSE,
 	SYSTEM_SUSPEND_INIT,
 	SYSTEM_SUSPEND_COMPLETE,
+	GUEST_ENTER_VM_OP,
 	MAX_FUNC_ID,
 };
 
@@ -130,4 +131,6 @@ typedef enum  {
 #define GUEST_PAUSE_CMD(vmid)		CREATE_CMD(GUEST_PAUSE, vmid)
 #define SYS_SUSPEND_INIT_CMD		CREATE_CMD(SYSTEM_SUSPEND_INIT, 0)
 #define SYS_SUSPEND_COMPLETE_CMD	CREATE_CMD(SYSTEM_SUSPEND_COMPLETE, 0)
+#define GUEST_ENTER_VM_OP_CMD		CREATE_CMD(GUEST_ENTER_VM_OP, 0)
+
 #endif /* _TEGRA_HV_SYSMGR_H */
-- 
2.43.0

